apps/wallet/src/engine/runners/solana/history.ts

```diff
@@ -0,0 +1,318 @@
+	) as ParsedInstruction;
+
+	const sender = (parsedInstruction?.parsed?.info?.source as string) || '';
+	const receiver =
+		(parsedInstruction?.parsed?.info?.destination as string) || '';
+	const amount =
+		((parsedInstruction?.parsed?.info?.lamports as number) || 0) / 10 ** 9;
+
+	return {
+		sender,
+		receiver,
+		amount,
+		token,
+	};
+};
+
+const getSplTransactionBalances = async (
+	connection: Connection,
+	meta: ParsedTransactionMeta,
+) => {
+	if (!meta.preTokenBalances || !meta.postTokenBalances) return;
+	if (meta.preTokenBalances.length === 0 || meta.postTokenBalances.length === 0)
+		return;
+
+	const { mint, sender, receiver, amount } = analyzeTokenBalances(
+		meta.preTokenBalances,
+		meta.postTokenBalances,
+	);
+
+	const token: TransactionHistory['token'] = {
+		network: Networks.solana,
+		metadata: await getMetadata(connection, mint),
+	};
+
+	return {
+		sender,
+		receiver,
+		amount: Math.abs(amount),
+		token,
+	};
+};
+
+const analyzeTokenBalances = (
+	preTokenBalances: TokenBalance[],
+	postTokenBalances: TokenBalance[],
+) => {
+	if (postTokenBalances.length > preTokenBalances.length) {
+		const referralPreBalance = preTokenBalances[0];
+		const correspondingPostBalance = postTokenBalances.find(
+			(balance) => balance.accountIndex === referralPreBalance.accountIndex,
+		);
+		const otherPostBalance = postTokenBalances.find(
+			(balance) => balance.accountIndex !== referralPreBalance.accountIndex,
+		);
+		const mint = referralPreBalance.mint;
+		const amount =
+			(correspondingPostBalance?.uiTokenAmount.uiAmount || 0) -
+			(referralPreBalance?.uiTokenAmount.uiAmount || 0);
+
+		let sender: string;
+		let receiver: string;
+		if (amount < 0) {
+			sender = correspondingPostBalance?.owner || '';
+			receiver = otherPostBalance?.owner || '';
+		} else {
+			sender = otherPostBalance?.owner || '';
+			receiver = correspondingPostBalance?.owner || '';
+		}
+
+		return {
+			mint,
+			amount,
+			sender,
+			receiver,
+		};
+	}
+
+	const referralPostBalance = postTokenBalances[0];
+	const correspondingPreBalance = preTokenBalances.find(
+		(balance) => balance.accountIndex === referralPostBalance.accountIndex,
+	);
+	const otherPreBalance = preTokenBalances.find(
+		(balance) => balance.accountIndex !== referralPostBalance.accountIndex,
+	);
+	const mint = referralPostBalance.mint;
+	const amount =
+		(referralPostBalance.uiTokenAmount.uiAmount || 0) -
+		(correspondingPreBalance?.uiTokenAmount.uiAmount || 0);
+
+	let sender: string;
+	let receiver: string;
+	if (amount < 0) {
+		sender = correspondingPreBalance?.owner || '';
+		receiver = otherPreBalance?.owner || '';
+	} else {
+		sender = otherPreBalance?.owner || '';
+		receiver = correspondingPreBalance?.owner || '';
+	}
+
+	return {
+		mint,
+		amount,
+		sender,
+		receiver,
+	};
+};
+
+const getTransactionFee = async (walletAddress: string) => {
+	return await storage.safeGet<TokenDocument>(
+		`${walletAddress}/token/${solana.solMint}`,
+	);
+};
```

Comment:
why is it named `getTransactionFee`, I saw it just gets the native token document

Reply:
this method means to get the transaction's tokenForFee, in the future, we might fetch other token using in Gasilon transactions. I renamed it to `getTransactionTokenFee` for more transparent meaning